tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp

```diff
@@ -64,15 +64,74 @@ void EditorDiagConsumer::getAllDiagnostics(
 
-  // Note: we cannot reuse InputBufIds because there may be diagnostics outside
-  // the inputs.  Instead, sort the extant buffers.
+  // Note: We cannot use the input buffer IDs because there may be diagnostics
+  // outside the inputs. Instead, sort the extant buffers.
   auto bufferIDs = getSortedBufferIDs(BufferDiagnostics);
   for (unsigned bufferID : bufferIDs) {
     const auto &diags = BufferDiagnostics[bufferID];
     Result.append(diags.begin(), diags.end());
   }
 }
 
+/// Retrieve the raw range from a range, if it exists in the provided buffer.
+/// Otherwise returns \c None.
+static Optional<RawCharSourceRange> getRawRangeInBuffer(CharSourceRange range,
+                                                        unsigned bufferID,
+                                                        SourceManager &SM) {
+  if (!range.isValid() || !SM.findBufferContainingLoc(range.getStart()))
+    return None;
+
+  unsigned offset = SM.getLocOffsetInBuffer(range.getStart(), bufferID);
+  unsigned length = range.getByteLength();
+  return {{offset, length}};
+}
+
+BufferInfoSharedPtr EditorDiagConsumer::getBufferInfo(
+    StringRef FileName, Optional<unsigned> BufferID, swift::SourceManager &SM) {
+  // NOTE: Using StringRef as a key here relies on SourceMgr using const char*
+  // as buffer identifiers. This is fast, but may be brittle.  We can always
+  // switch over to using a StringMap. Note that the logic in
+  // SerializedDiagnosticConsumer::getEmitFile will also need changing.
+  auto Result = BufferInfos.find(FileName);
+  if (Result != BufferInfos.end())
+    return Result->second;
+
+  Optional<std::string> GeneratedFileText;
+  Optional<BufferInfo::OriginalLocation> OriginalLocInfo;
+
+  // If we have a generated buffer, we need to include the source text, as
+  // clients otherwise won't be able to access to it.
+  if (BufferID) {
+    if (auto Info = SM.getGeneratedSourceInfo(*BufferID)) {
+      // Don't include this information for replaced function body buffers, as
+      // they'll just be referencing the original file.
+      // FIXME: Does this mean that the location information we'll be giving
+      // will be wrong? Seems like we may need to adjust it to correspond to the
+      // original file.
+      if (Info->kind != GeneratedSourceInfo::ReplacedFunctionBody) {
+        GeneratedFileText.emplace(SM.getEntireTextForBuffer(*BufferID));
+
+        auto OrigRange = Info->originalSourceRange;
+        if (OrigRange.isValid()) {
+          auto OrigLoc = OrigRange.getStart();
+          auto OrigBuffer = SM.findBufferContainingLoc(OrigLoc);
+
+          auto OrigRawRange = getRawRangeInBuffer(OrigRange, OrigBuffer, SM);
+          assert(OrigRawRange);
+
+          auto OrigFileName = SM.getDisplayNameForLoc(OrigLoc);
+          auto OrigInfo = getBufferInfo(OrigFileName, OrigBuffer, SM);
+          OriginalLocInfo.emplace(OrigInfo, OrigBuffer, *OrigRawRange);
```

Comment:
This recursion assumes original locations forms a DAG, is that a safe assumption to make? We could eagerly create and insert the buffer info if we want to be defensive here.

Reply:
I can't imagine a case that it wouldn't be, seems fine to me.